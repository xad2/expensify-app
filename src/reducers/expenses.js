// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      console.log('ADD', action.expense);
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      console.log('REMOVE', action.expense);
      return state.filter( (expense) => 
        expense.id !== action.id 
      );
    case 'EDIT_EXPENSE':
      return state.map( (expense) => {
        if( expense.id === action.id ) {
          console.log('EDIT', action.updates);
          return {
            ...expense,
            ...action.updates // spread object
          };
        } else {
          return expense;
        };
      });
    default: return state;
  }
};

export default expensesReducer;