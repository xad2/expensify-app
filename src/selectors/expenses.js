import moment from 'moment';

// Get visible expenses (filter)
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense, index) => {
    const createdAtMoment = moment(expense.createdAt);
    
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;

    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    const textMatch = expense.description.toLowerCase().trim().includes(text.toLowerCase().trim());


    return startDateMatch && endDateMatch && textMatch;
  }).sort((a,b) => { // if return is -1 put a before b, else if a return 1 put b before a, else return same 
    if (sortBy === 'date'){ // if a < b return 1, else a > b return 1, else a = b return 0
      return a.createdAt < b.createdAt? 1 : -1; // from latest 
    }else if( sortBy === 'amount'){
      return a.amount < b.amount? 1 : -1; // from greatest to lowest
    }
  });
};

export default getVisibleExpenses;