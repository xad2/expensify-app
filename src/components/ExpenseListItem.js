import React from 'react';
// import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import {Link, NavLink} from 'react-router-dom';

export default (
  ({expense}) => {
  // (props) => {
    return (
      <div>
        <div><Link to={`/edit/${expense.id}`} >{expense.description}</Link></div>
        <div>{expense.amount}</div>
        <div>{expense.createdAt}</div>        
      </div>
    );
  }
);

 