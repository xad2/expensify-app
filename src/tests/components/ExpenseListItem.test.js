import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../dummydata/expenses';

test('should render ExpenseListItem with dummy data', () => {
  const wrapper = shallow(<ExpenseListItem key={expenses[0].id} expense={expenses[0]}/>);
  expect(wrapper).toMatchSnapshot();
});