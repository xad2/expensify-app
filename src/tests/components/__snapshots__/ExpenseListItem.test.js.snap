// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with dummy data 1`] = `
<div>
  <div>
    <Link
      replace={false}
      to="/edit/1"
    >
      Gun
    </Link>
  </div>
  <div>
    194
  </div>
  <div>
    0
  </div>
</div>
`;
